### ISSUE TITLE
Semantics of Unix.exec* function on Windows

### ISSUE THREAD

**Issue by yakobowski** [ocaml/ocaml#11754]
We just got bitten (the hard way, as the proverbial saying goes) by the semantics of the `Unix.exec*` functions on Windows. Interleaved calls to a binary (partly) executing OCaml code, and to `echo`, suddenly resulted in something that did not respect the sequencing in our script shell. That is

```
echo Step 1
<call to binary>
echo Step 2
<another call to binary>
```

produced
```
Step 1
Step 2
<output for the first call>
<output for the second call>
```

After investigation, I saw that those functions are calling _wexecv [1]. More searching resulted in [2], which I quote below:

> Using the _wexec* family of functions (or _P_OVERLAY mode with _wspawn*) in Windows is generally a bad idea, especially for console applications (the default link target for the [w]main entry point). NT has no equivalent to the exec* family implemented for Windows processes, so the CRT simply spawns a new process and exits the current process. If a console-based shell is waiting on the current process, it will resume its standard I/O REPL, and now we have a mess on our hands, with two processes writing to the console and competing for access to console input.

This perfectly matches our symptoms, and replacing our use of `exec*` by calls to `CreateProcess` fixed everything. Based on the Stack Overflow comment, I doubt that the Windows implementation of the `exec*` functions can be made to match the Unix behavior. But some words of warning would help. If there is a consensus, I will propose a PR on `unix.mli` to that effect.

[1] https://github.com/ocaml/ocaml/blob/4f231696a871720cfd279713c1bfa8f6a9fa80c2/runtime/caml/misc.h#L379
[2] https://stackoverflow.com/questions/53218984/how-can-i-avoid-the-windows-exec-and-spawn-functions-splitting-arguments-at#comment93349909_53218984

**Comment by nojb** [comment]
> But some words of warning would help. If there is a consensus, I will propose a PR on `unix.mli` to that effect.

I think this is a good idea. Thanks!

**Comment by nojb** [comment]
Friendly ping.

**Comment by dbuenzli** [comment]
As I [mentioned](https://github.com/ocaml/ocaml/pull/11892#discussion_r1070565463) in the doc improvement PR. I think the current Windows implementation is more a curse than a blessing and I'm not sure it's really possible to build something cross-platform on top of it.

In my opinion it would be better to either:

1. Emulate more of the semantics of it on Windows e.g. see [here](https://github.com/b0-system/b0/blob/152c47864e9da478027bfa1113b892334abecffc/src/std/b0_std.ml#L3936-L3960) -- which is, as @yakobowski mentioned, not without caveats either. 
2. Simply let the function error with `ENOSYS` on windows.

**Comment by gasche** [comment]
(cc also @dra27 @xavierleroy @MisterDA who may have an opinion on this )

**Comment by nojb** [comment]
Interesting Hakell PR about the same issue (along the lines of 1. above): https://github.com/haskell/process/pull/80

**Comment by Octachron** [comment]
Fixed (for the documentation part) in #11892 .
